{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "library(Biobase)\n",
    "library(pcaMethods)\n",
    "library(colorspace)\n",
    "library(RColorBrewer)\n",
    "library(sva)\n",
    "library(DASC)\n",
    "library(MASS)\n",
    "library(NMF)\n",
    "library(dendextend)\n",
    "library(amap)\n",
    "library(pvclust)\n",
    "library(Rtsne)\n",
    "library(vioplot)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "date_sra = '2018_5_1'\n",
    "dir_ensembl = '/Users/kf/Dropbox/kfdata/02_Data/my_db/Ensembl/release-91/'\n",
    "dir_curated_transcriptome = paste0(dir_ensembl, 'curated_transcriptome/', date_sra, '/')\n",
    "dir_tc = paste0(dir_curated_transcriptome, 'tc/')\n",
    "dir_sra = paste0(dir_curated_transcriptome, 'sra/')\n",
    "dir_sva = paste0(dir_curated_transcriptome, 'sva/')\n",
    "#species = 'Anolis_carolinensis'\n",
    "#infile = paste0(dir_ensembl, 'kallisto_summary/', date_sra, '/tpm.masked.kallisto.gene.log.tsv/', species, '.gene.log.tsv')\n",
    "#srafile = paste0(dir_ensembl, 'sra/', date_sra, '/sra_table_mapped_', date_sra, '.tsv')\n",
    "outdir = '/Users/kf/Dropbox/kfdata/02_Data/04_Convergence_Duplication/20180207_kallisto/'\n",
    "dist_method=\"pearson\"\n",
    "min_dif = 0\n",
    "selected_tissues = strsplit('brain|heart|kidney|liver|ovary|testis', '\\\\|')[[1]]\n",
    "\n",
    "if (!endsWith(outdir, '/')) {\n",
    "    outdir = paste0(outdir, '/')\n",
    "}\n",
    "\n",
    "# set directory\n",
    "if (!file.exists(outdir)) {\n",
    "    dir.create(outdir)\n",
    "}\n",
    "setwd(outdir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "tc_sra_intersect = function(tc, sra) {\n",
    "    sra_run = sra$run\n",
    "    tc = tc[,colnames(tc) %in% sra_run]\n",
    "    sra = sra[sra$run %in% colnames(tc), ]\n",
    "    return(list(tc=tc, sra=sra))\n",
    "}\n",
    "\n",
    "remove_nonexpressed_gene = function(tc) {\n",
    "    gene_sum = apply(tc, 1, sum)\n",
    "    tc_ex = tc[gene_sum > 0,]\n",
    "    tc_ne = tc[gene_sum == 0,]\n",
    "    return(list(tc_ex=tc_ex, tc_ne=tc_ne))\n",
    "}\n",
    "\n",
    "add_color_to_sra = function(sra, selected_tissues) {\n",
    "    sra = sra[,(!names(sra) %in% c('bp_color','sp_color','tissue_color'))]\n",
    "    bioproject = as.character(sra$bioproject)\n",
    "    scientific_name = as.character(sra$scientific_name)\n",
    "    tissue = as.character(sra$tissue)\n",
    "    if (length(selected_tissues) <= 8) {\n",
    "        tissue_color = brewer.pal(length(unique(tissue)), \"Dark2\")\n",
    "        bp_color = rainbow_hcl(length(unique(bioproject)), c=50)\n",
    "        sp_color = rainbow_hcl(length(unique(scientific_name)), c=100)\n",
    "    } else if (length(selected_tissues) <= 12) {\n",
    "        tissue_color = brewer.pal(length(unique(tissue)), \"Paired\")\n",
    "        bp_color = rainbow_hcl(length(unique(bioproject)), c=50)\n",
    "        sp_color = rainbow_hcl(length(unique(scientific_name)), c=100)\n",
    "    } else {\n",
    "        tissue_color = rainbow_hcl(length(selected_tissues), c=100)\n",
    "        bp_color = rainbow_hcl(length(unique(bioproject)), c=50)\n",
    "        sp_color = rainbow_hcl(length(unique(scientific_name)), c=150)\n",
    "    }\n",
    "    df_tissue = data.frame(tissue=sort(unique(tissue)), tissue_color=tissue_color[1:length(sort(unique(tissue)))], stringsAsFactors=FALSE)\n",
    "    df_bp = data.frame(bioproject=sort(unique(bioproject)), bp_color=bp_color[1:length(sort(unique(bioproject)))], stringsAsFactors=FALSE)\n",
    "    df_sp = data.frame(scientific_name=sort(unique(scientific_name)), sp_color=sp_color[1:length(sort(unique(scientific_name)))], stringsAsFactors=FALSE)\n",
    "    sra = merge(sra, df_bp, sort=FALSE, all.y=FALSE)\n",
    "    sra = merge(sra, df_sp, sort=FALSE, all.y=FALSE)\n",
    "    sra = merge(sra, df_tissue, sort=FALSE, all.y=FALSE)\n",
    "    return(sra)\n",
    "}\n",
    "\n",
    "sort_tc_and_sra = function(tc, sra, sort_columns=c(\"tissue\",\"scientific_name\",\"bioproject\")) {\n",
    "  for (column in rev(sort_columns)) {\n",
    "    sra = sra[order(sra[column]),]\n",
    "  }\n",
    "  tc = tc[,sra$run[sra$run %in% colnames(tc)]]\n",
    "  return(list(tc=tc, sra=sra))\n",
    "}\n",
    "\n",
    "sort_averaged_tc = function(tc) {\n",
    "    split_colnames = strsplit(colnames(tc), \"_\")\n",
    "    genus_names = c()\n",
    "    specific_names = c()\n",
    "    tissue_names = c()\n",
    "    for (i in 1:length(split_colnames)) {\n",
    "        genus_names = c(genus_names, split_colnames[[i]][1])\n",
    "        specific_names = c(specific_names, split_colnames[[i]][2])\n",
    "        tissue_names = c(tissue_names, split_colnames[[i]][3])\n",
    "    }\n",
    "    colname_order = order(tissue_names, genus_names, specific_names)\n",
    "    tc = tc[, colname_order]\n",
    "    return(tc)\n",
    "}\n",
    "\n",
    "cleanY = function(y, mod, svs) {\n",
    "  X = cbind(mod, svs)\n",
    "  Hat = solve(t(X) %*% X) %*% t(X)\n",
    "  beta = (Hat %*% t(y))\n",
    "  P = ncol(mod)\n",
    "  return(y - t(as.matrix(X[,-c(1:P)]) %*% beta[-c(1:P),]))\n",
    "}\n",
    "\n",
    "map_color = function(redundant_variables, c) {\n",
    "    uniq_var = unique(redundant_variables)\n",
    "    uniq_col = rainbow_hcl(length(uniq_var), c=c)\n",
    "    df_unique = data.frame(var=uniq_var, col=uniq_col, stringsAsFactors=FALSE)\n",
    "    df_redundant = data.frame(var=redundant_variables, order=seq(1, length(redundant_variables)), stringsAsFactors=FALSE)\n",
    "    df_redundant = merge(df_redundant, df_unique, by=\"var\", all.x=TRUE, stringsAsFactors=FALSE)\n",
    "    df_redundant = df_redundant[order(df_redundant$order),]\n",
    "    return(df_redundant$col)\n",
    "}\n",
    "\n",
    "color_children2parent = function(node) {\n",
    "    if (length(node)==2) {\n",
    "        child1_color = attributes(node[[1]])$edgePar[['col']]\n",
    "        child2_color = attributes(node[[2]])$edgePar[['col']]\n",
    "        if ((!is.null(child1_color))&(!is.null(child2_color))) {\n",
    "            if (child1_color==child2_color) {\n",
    "                attributes(node)$edgePar[['col']] = child1_color\n",
    "            }\n",
    "        }\n",
    "    }\n",
    "    return(node)\n",
    "}\n",
    "\n",
    "draw_dendrogram = function(sra, tc_dist_dist, fontsize=7) {\n",
    "    dend <- as.dendrogram(hclust(tc_dist_dist))\n",
    "    dend_colors = sra$tissue_color[order.dendrogram(dend)]\n",
    "    labels_colors(dend) <- dend_colors\n",
    "    dend_labels <- sra$run[order.dendrogram(dend)]\n",
    "    dend <- color_branches(dend, labels=dend_labels, col=dend_colors)\n",
    "    dend <- set(dend, \"branches_lwd\", 1)\n",
    "    for (i in 1:ncol(tc)) {\n",
    "        dend = dendrapply(dend, color_children2parent)\n",
    "    }\n",
    "    cex.xlab = min(fontsize, max(0.2, 0.5/log10(nrow(sra))))\n",
    "    par(cex=cex.xlab)\n",
    "    plot(dend, las=1, axes=FALSE)\n",
    "    par(cex=1)\n",
    "    axis(side=2, line=0, las=1)\n",
    "    mtext('Distance', side=2, line=8.5, outer=FALSE)\n",
    "    n = nrow(sra)\n",
    "    symbols(1:n, rep(0,n), circles=rep(1,n), add=TRUE, inches=0.02, xpd=TRUE, lwd=1, \n",
    "            bg=sra$tissue_color[order.dendrogram(dend)], fg=sra$bp_color[order.dendrogram(dend)])\n",
    "}\n",
    "\n",
    "draw_pca = function(sra, tc_dist_matrix, fontsize=7) {\n",
    "    set.seed(1)\n",
    "    pca = prcomp(tc_dist_matrix)\n",
    "    xlabel = paste0(\"PC 1 (\", round(summary(pca)$importance[2,1]*100, digits=1), \"%)\")\n",
    "    ylabel = paste0(\"PC 2 (\", round(summary(pca)$importance[2,2]*100, digits=1), \"%)\")\n",
    "    plot(pca$rotation[,1], pca$rotation[,2], pch=21, cex=2, lwd=1, bg=sra$tissue_color, col=sra$bp_color, xlab=xlabel, ylab=ylabel, las=1)\n",
    "    #plot(pca$x[,1], pca$x[,2], pch=21, cex=2, lwd=2, bg=sra$tissue_color, col=sra$bp_color, main=title, xlab=xlabel, ylab=ylabel, las=1)\n",
    "}\n",
    "\n",
    "draw_mds = function(sra, tc_dist_dist, fontsize=7) {\n",
    "    set.seed(1)\n",
    "    try_out = tryCatch(\n",
    "        {isoMDS(tc_dist_dist, k=2, maxit=100)},\n",
    "        error = function(a){return(\"MDS failed.\")}\n",
    "    )\n",
    "    if (mode(try_out)==\"character\") {\n",
    "        cat('MDS failed.\\n')\n",
    "        plot(c(0, 1), c(0, 1), ann = F, bty = 'n', type = 'n', xaxt = 'n', yaxt = 'n')\n",
    "    } else {\n",
    "        mds <- try_out\n",
    "        plot(mds$points[,1], mds$points[,2], pch=21, cex=2, lwd=1, bg=sra$tissue_color, col=sra$bp_color, xlab=\"MDS dimension 1\", ylab=\"MDS dimension 2\", las=1)\n",
    "    }\n",
    "}\n",
    "\n",
    "draw_tau_histogram = function(tc, sra, selected_tissues, fontsize=7) {\n",
    "    df_tau = tissue2tau(tissue_mean(tc, sra, selected_tissues), rich.annotation=FALSE, unlog=TRUE)\n",
    "    hist_out = hist(df_tau$tau, breaks=seq(0,1,0.05), las=1, xlab='Tau (expression specificity)', ylab='Gene count', main='', col='gray')\n",
    "    num_noexp = sum(is.na(df_tau$tau))\n",
    "    num_all = nrow(df_tau)\n",
    "    #num_exp = nrow(df_tau) - num_noexp\n",
    "    #text_noexp = paste('Expressed genes:', num_exp, '\\nNon-expressed genes:', num_noexp)\n",
    "    text_noexp = paste0('Excluded due to\\nno expression:\\n', num_noexp, '/', num_all, ' genes')\n",
    "    text(0, max(hist_out$counts)*0.85, text_noexp, pos=4)\n",
    "}\n",
    "\n",
    "draw_exp_level_histogram = function(tc, sra, selected_tissues, fontsize=7) {\n",
    "    tc_tissue = tissue_mean(tc, sra, selected_tissues)\n",
    "    xmax = apply(tc_tissue, 1, max)\n",
    "    xmax[xmax<0] = 0\n",
    "    xmax[xmax>15] = 15\n",
    "    breaks = seq(0, 15, 1)\n",
    "    hist_out = hist(xmax, breaks=breaks, las=1, xlab='Max expression (log TPM+1)', ylab='Gene count', main='', col='gray')\n",
    "}\n",
    "\n",
    "draw_legend = function(sra, new=TRUE, pos=\"center\", fontsize=7, nlabel.in.col) {\n",
    "    if (new) {\n",
    "        plot.new()\n",
    "    }\n",
    "    tissue_unique = unique(sra$tissue)\n",
    "    bp_unique = unique(sub(';.*', '', sra$bioproject))\n",
    "    tissue_color_unique = unique(sra$tissue_color)\n",
    "    bp_color_unique = unique(sra$bp_color)\n",
    "    ncol = ceiling((length(tissue_unique)+length(bp_unique)+2)/nlabel.in.col)\n",
    "    legend_text = c('Organ', as.character(tissue_unique), '', 'BioProject',as.character(bp_unique))\n",
    "    legend_color = c(rgb(1,1,1,0), rep(rgb(1,1,1,0), length(tissue_color_unique)), rgb(1,1,1,0), rgb(1,1,1,0), bp_color_unique)\n",
    "    legend_bg = c(rgb(1,1,1,0), tissue_color_unique, rgb(1,1,1,0), rgb(1,1,1,0), rep(rgb(1,1,1,0), length(bp_color_unique)))\n",
    "    legend_font = c(2, rep(1, length(tissue_color_unique)), 1, 2, rep(1, length(bp_color_unique)))\n",
    "    legend(pos, legend=legend_text, pch=21, lwd=1, lty=0, col=legend_color, pt.bg=legend_bg, text.font=legend_font, ncol=ncol, bty='n')\n",
    "}\n",
    "\n",
    "save_plot = function(tc, sra, sva_out, dist_method, file, selected_tissues, fontsize=7) {\n",
    "    out = tc_sra_intersect(tc, sra) ; tc = out[['tc']] ; sra = out[['sra']]\n",
    "    sra = add_color_to_sra(sra, selected_tissues)\n",
    "    out = sort_tc_and_sra(tc, sra) ; tc = out[[\"tc\"]] ; sra = out[[\"sra\"]]\n",
    "    pdf(paste0(file, \".pdf\"), height=8, width=7.2, fonts='Helvetica', pointsize=fontsize)\n",
    "    layout_matrix=matrix(c(\n",
    "        2,2,2,2,2,1,1,1,1,1,1,1,\n",
    "        2,2,2,2,2,1,1,1,1,1,1,1,\n",
    "        2,2,2,2,2,1,1,1,1,1,1,1,\n",
    "        2,2,2,2,2,1,1,1,1,1,1,1,\n",
    "        2,2,2,2,2,1,1,1,1,1,1,1,\n",
    "        3,3,3,4,4,4,5,5,5,6,6,6,\n",
    "        3,3,3,4,4,4,5,5,5,6,6,6,\n",
    "        3,3,3,4,4,4,5,5,5,6,6,6,\n",
    "        7,7,7,8,8,8,9,9,9,9,9,9,\n",
    "        7,7,7,8,8,8,9,9,9,9,9,9,\n",
    "        7,7,7,8,8,8,9,9,9,9,9,9,\n",
    "        7,7,7,8,8,8,9,9,9,9,9,9,\n",
    "        10,10,10,10,10,10,10,10,10,10,10,10,\n",
    "        10,10,10,10,10,10,10,10,10,10,10,10\n",
    "        ),\n",
    "      14, 12, byrow=TRUE\n",
    "    )\n",
    "    layout(layout_matrix)\n",
    "    tc_dist_matrix = cor(tc, method=dist_method)\n",
    "    tc_dist_matrix[is.na(tc_dist_matrix)] = 0\n",
    "    tc_dist_dist = Dist(t(tc), method=dist_method) + 0.000000001\n",
    "    tc_dist_dist[is.na(tc_dist_dist)] = 1\n",
    "    par(mar=c(6,6,1,0)); draw_dendrogram(sra, tc_dist_dist, fontsize)\n",
    "    par(mar=c(0,0,0,0)); draw_heatmap(sra, tc_dist_matrix, legend=FALSE)\n",
    "    #draw_dendrogram(sra, tc, nboot=1000, cex.xlab=0.6, pvclust_file=paste0(file, '.pvclust.RData'))\n",
    "    par(mar=c(4,4,0.1,1)); draw_pca(sra, tc_dist_matrix, fontsize)\n",
    "    par(mar=c(4,4,0.1,1)); draw_mds(sra, tc_dist_dist, fontsize)\n",
    "    par(mar=c(4,4,0.1,1)); draw_tsne(sra, tc, fontsize)\n",
    "    par(mar=c(4,5,0.1,1)); draw_boxplot(sra, tc_dist_matrix, fontsize)\n",
    "    par(mar=c(4,4,1,1)); draw_exp_level_histogram(tc, sra, selected_tissues, fontsize)\n",
    "    par(mar=c(4,4,1,1)); draw_tau_histogram(tc, sra, selected_tissues, fontsize)\n",
    "    par(mar=rep(0.1,4)); df_r2 = draw_sva_summary(sva_out, tc, sra, fontsize)\n",
    "    if (! all(is.na(df_r2))) {\n",
    "        write.table(df_r2, paste0(file,'.r2.tsv'), sep='\\t', row.names=FALSE)\n",
    "    }\n",
    "    par(mar=rep(0.1,4)); draw_legend(sra, new=TRUE, pos=\"center\", fontsize=fontsize, nlabel.in.col=8)\n",
    "    graphics.off()\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Anolis carolinensis \n",
      "Astyanax mexicanus \n",
      "Bos taurus \n",
      "Callithrix jacchus \n",
      "Canis lupus \n",
      "Chinchilla lanigera \n",
      "Danio rerio \n",
      "Gadus morhua \n",
      "Gallus gallus \n",
      "Homo sapiens \n",
      "Macaca mulatta \n",
      "Monodelphis domestica \n",
      "Mus musculus \n",
      "Oreochromis niloticus \n",
      "Ornithorhynchus anatinus \n",
      "Oryctolagus cuniculus \n",
      "Oryzias latipes \n",
      "Ovis aries \n",
      "Rattus norvegicus \n",
      "Sus scrofa \n",
      "Xenopus tropicalis \n",
      "done.\n"
     ]
    }
   ],
   "source": [
    "# heatmap\n",
    "\n",
    "draw_heatmap2 = function(sra, tc_dist_matrix, legend=TRUE, cbar=TRUE, main=NA, fontsize=7) {\n",
    "    bp_fac = factor(sub(';.*', '', sra[,c(\"bioproject\")]))\n",
    "    tissue_fac = factor(sra[,c(\"tissue\")])\n",
    "    ann_label = data.frame(bioproject=bp_fac, tissue=tissue_fac)\n",
    "    bp_col_uniq = unique(sra$bp_color[order(sra$bioproject)])\n",
    "    tissue_col_uniq = unique(sra$tissue_color[order(sra$tissue)])\n",
    "    ann_color = list(bioproject=bp_col_uniq, tissue=tissue_col_uniq)\n",
    "    breaks = c(0, seq(0.3, 1, 0.01))\n",
    "    aheatmap(tc_dist_matrix, color=\"-RdYlBu2:71\", main=main, labRow=NA, labCol=NA,\n",
    "             Rowv=NA, Colv=NA, revC=TRUE, legend=cbar, breaks=breaks, \n",
    "             annCol=ann_label, annRow=ann_label, annColors=ann_color, annLegend=legend, fontsize=fontsize)\n",
    "}\n",
    "\n",
    "fontsize = 6\n",
    "spp = c('Bos taurus','Homo sapiens','Gallus gallus','Macaca mulatta','Canis lupus','Sus scrofa','Ovis aries','Mus musculus','Callithrix jacchus',\n",
    "'Rattus norvegicus','Monodelphis domestica','Oreochromis niloticus','Oryctolagus cuniculus','Danio rerio','Anolis carolinensis','Chinchilla lanigera',\n",
    "'Astyanax mexicanus','Xenopus tropicalis','Oryzias latipes','Gadus morhua','Ornithorhynchus anatinus')\n",
    "spp = spp[order(spp)]\n",
    "\n",
    "pdf('paneled_species_heatmap.pdf', height=6.5, width=7.2, fonts='Helvetica', pointsize=fontsize)\n",
    "par(mfrow=c(4,6), mar=c(0,0,0,0))\n",
    "\n",
    "for (sp in spp) {\n",
    "    cat(sp, '\\n')\n",
    "    sp_filled = sub(' ', '_', sp)\n",
    "    \n",
    "    files = list.files(dir_sra)\n",
    "    file = files[grep(sp_filled, files)]\n",
    "    sra = read.table(paste0(dir_sra, file), sep='\\t', header=TRUE, comment.char='', check.names=FALSE, quote='', stringsAsFactors=FALSE)\n",
    "    \n",
    "    files = list.files(dir_tc)\n",
    "    file = files[grep(sp_filled, files)]\n",
    "    tc = read.table(paste0(dir_tc, file), sep='\\t', header=TRUE, quote='', stringsAsFactors=FALSE)\n",
    "\n",
    "    sra = add_color_to_sra(sra, selected_tissues)\n",
    "    out = tc_sra_intersect(tc, sra) ; tc = out[['tc']] ; sra = out[['sra']]\n",
    "    out = sort_tc_and_sra(tc, sra) ; tc = out[[\"tc\"]] ; sra = out[[\"sra\"]]\n",
    "\n",
    "    tc_dist_matrix = cor(tc, method=dist_method)\n",
    "    tc_dist_matrix[is.na(tc_dist_matrix)] = 0\n",
    "    \n",
    "    main = substitute(italic(x), list(x=sp))\n",
    "    \n",
    "    draw_heatmap2(sra, tc_dist_matrix, legend=FALSE, cbar=FALSE, main=main, fontsize=fontsize)\n",
    "    \n",
    "    #if (sp=='Bos taurus') {\n",
    "    #    break\n",
    "    #}\n",
    "}\n",
    "graphics.off()\n",
    "cat('done.\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Anolis carolinensis \n",
      "Astyanax mexicanus \n",
      "Bos taurus \n",
      "Callithrix jacchus \n",
      "Canis lupus \n",
      "Chinchilla lanigera \n",
      "Danio rerio \n"
     ]
    }
   ],
   "source": [
    "# t-SNE\n",
    "\n",
    "draw_tsne2 = function(sra, tc, fontsize=7, main=NA) {\n",
    "    perplexity = min(30, floor(nrow(sra)/4))\n",
    "    set.seed(1)\n",
    "    out_tsne = Rtsne(as.matrix(t(tc)), theta=0, check_duplicates=FALSE, verbose=FALSE, perplexity=perplexity, dims=2)\n",
    "    try_out = tryCatch(\n",
    "        {\n",
    "            plot(out_tsne$Y[,1], out_tsne$Y[,2], pch=21, cex=2, lwd=1, bg=sra$tissue_color, col=sra$bp_color, \n",
    "                 xaxt='n', yaxt='n', las=1, main=main)\n",
    "        },\n",
    "        error = function(a){return(\"t-SNE plot failed.\")}\n",
    "    )\n",
    "    if (mode(try_out)==\"character\") {\n",
    "        cat('t-SNE failed.\\n')\n",
    "        plot(c(0, 1), c(0, 1), ann = F, bty = 'n', type = 'n', xaxt = 'n', yaxt = 'n')\n",
    "    }    \n",
    "}\n",
    "\n",
    "fontsize = 8\n",
    "spp = c('Bos taurus','Homo sapiens','Gallus gallus','Macaca mulatta','Canis lupus','Sus scrofa','Ovis aries','Mus musculus','Callithrix jacchus',\n",
    "'Rattus norvegicus','Monodelphis domestica','Oreochromis niloticus','Oryctolagus cuniculus','Danio rerio','Anolis carolinensis','Chinchilla lanigera',\n",
    "'Astyanax mexicanus','Xenopus tropicalis','Oryzias latipes','Gadus morhua','Ornithorhynchus anatinus')\n",
    "spp = spp[order(spp)]\n",
    "\n",
    "pdf('paneled_species_t-SNE.pdf', height=5.5, width=7.2, fonts='Helvetica', pointsize=fontsize)\n",
    "par(mfrow=c(4,6), mar=c(0,0,0,0))\n",
    "\n",
    "for (sp in spp) {\n",
    "    cat(sp, '\\n')\n",
    "    sp_filled = sub(' ', '_', sp)\n",
    "    \n",
    "    files = list.files(dir_sra)\n",
    "    file = files[grep(sp_filled, files)]\n",
    "    sra = read.table(paste0(dir_sra, file), sep='\\t', header=TRUE, comment.char='', check.names=FALSE, quote='', stringsAsFactors=FALSE)\n",
    "    \n",
    "    files = list.files(dir_tc)\n",
    "    file = files[grep(sp_filled, files)]\n",
    "    tc = read.table(paste0(dir_tc, file), sep='\\t', header=TRUE, quote='', stringsAsFactors=FALSE)\n",
    "\n",
    "    sra = add_color_to_sra(sra, selected_tissues)\n",
    "    out = tc_sra_intersect(tc, sra) ; tc = out[['tc']] ; sra = out[['sra']]\n",
    "    out = sort_tc_and_sra(tc, sra) ; tc = out[[\"tc\"]] ; sra = out[[\"sra\"]]\n",
    "    \n",
    "    main = substitute(italic(x), list(x=sp))\n",
    "    \n",
    "    par(mar=c(0.2,0.2,2,0.2)); draw_tsne2(sra=sra, tc=tc, main=main, fontsize=fontsize)\n",
    "    \n",
    "    #if (sp=='Bos taurus') {\n",
    "    #    break\n",
    "    #}\n",
    "}\n",
    "graphics.off()\n",
    "cat('done.\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 158,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Anolis carolinensis \n",
      "Astyanax mexicanus \n",
      "Bos taurus \n",
      "Callithrix jacchus \n",
      "Canis lupus \n",
      "Chinchilla lanigera \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in min(x):\n",
      "“no non-missing arguments to min; returning Inf”Warning message in max(x):\n",
      "“no non-missing arguments to max; returning -Inf”Warning message in min(x):\n",
      "“no non-missing arguments to min; returning Inf”Warning message in max(x):\n",
      "“no non-missing arguments to max; returning -Inf”Warning message in min(x):\n",
      "“no non-missing arguments to min; returning Inf”Warning message in max(x):\n",
      "“no non-missing arguments to max; returning -Inf”"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Danio rerio \n",
      "Gadus morhua \n",
      "Gallus gallus \n",
      "Homo sapiens \n",
      "Macaca mulatta \n",
      "Monodelphis domestica \n",
      "Mus musculus \n",
      "Oreochromis niloticus \n",
      "Ornithorhynchus anatinus \n",
      "Oryctolagus cuniculus \n",
      "Oryzias latipes \n",
      "Ovis aries \n",
      "Rattus norvegicus \n",
      "Sus scrofa \n",
      "Xenopus tropicalis \n",
      "done.\n"
     ]
    }
   ],
   "source": [
    "# correlation boxplot\n",
    "\n",
    "draw_boxplot2 = function(sra, tc_dist_matrix, main, fontsize=7) {\n",
    "    is_same_bp = outer(sra$bioproject, sra$bioproject, function(x,y){x==y})\n",
    "    is_same_tissue = outer(sra$tissue, sra$tissue, function(x,y){x==y})\n",
    "    is_same_entry = outer(1:nrow(sra), 1:nrow(sra), function(x,y){x==y})\n",
    "    plot(c(0.5, 4.5), c(0, 1), main=main, type = 'n', xlab='', ylab=\"Pearson's correlation\\ncoefficient\", las=1, xaxt='n')\n",
    "    boxplot(tc_dist_matrix[(!is_same_bp)&(!is_same_tissue)&(!is_same_entry)], at=1, add=TRUE, col='gray', yaxt='n')\n",
    "    boxplot(tc_dist_matrix[(is_same_bp)&(!is_same_tissue)&(!is_same_entry)], at=2, add=TRUE, col='gray', yaxt='n')\n",
    "    boxplot(tc_dist_matrix[(!is_same_bp)&(is_same_tissue)&(!is_same_entry)], at=3, add=TRUE, col='gray', yaxt='n')\n",
    "    boxplot(tc_dist_matrix[(is_same_bp)&(is_same_tissue)&(!is_same_entry)], at=4, add=TRUE, col='gray', yaxt='n')\n",
    "    labels = c('bw\\nbw', 'bw\\nwi', 'wi\\nbw', 'wi\\nwi')\n",
    "    axis(side=1, at=c(1,2,3,4), labels=labels, padj=0.5)\n",
    "    axis(side=1, at=0.35, labels='Organ\\nBioProject', padj=0.5, hadj=1, tick=FALSE)\n",
    "}\n",
    "\n",
    "fontsize = 8\n",
    "spp = c('Bos taurus','Homo sapiens','Gallus gallus','Macaca mulatta','Canis lupus','Sus scrofa','Ovis aries','Mus musculus','Callithrix jacchus',\n",
    "'Rattus norvegicus','Monodelphis domestica','Oreochromis niloticus','Oryctolagus cuniculus','Danio rerio','Anolis carolinensis','Chinchilla lanigera',\n",
    "'Astyanax mexicanus','Xenopus tropicalis','Oryzias latipes','Gadus morhua','Ornithorhynchus anatinus')\n",
    "spp = spp[order(spp)]\n",
    "\n",
    "pdf('paneled_species_boxplot.pdf', height=5.5, width=7.2, fonts='Helvetica', pointsize=fontsize)\n",
    "par(mfrow=c(4,6), mar=c(0,0,0,0))\n",
    "\n",
    "for (sp in spp) {\n",
    "    cat(sp, '\\n')\n",
    "    sp_filled = sub(' ', '_', sp)\n",
    "    \n",
    "    files = list.files(dir_sra)\n",
    "    file = files[grep(sp_filled, files)]\n",
    "    sra = read.table(paste0(dir_sra, file), sep='\\t', header=TRUE, comment.char='', check.names=FALSE, quote='', stringsAsFactors=FALSE)\n",
    "    \n",
    "    files = list.files(dir_tc)\n",
    "    file = files[grep(sp_filled, files)]\n",
    "    tc = read.table(paste0(dir_tc, file), sep='\\t', header=TRUE, quote='', stringsAsFactors=FALSE)\n",
    "\n",
    "    sra = add_color_to_sra(sra, selected_tissues)\n",
    "    out = tc_sra_intersect(tc, sra) ; tc = out[['tc']] ; sra = out[['sra']]\n",
    "    out = sort_tc_and_sra(tc, sra) ; tc = out[[\"tc\"]] ; sra = out[[\"sra\"]]\n",
    "\n",
    "    tc_dist_matrix = cor(tc, method=dist_method)\n",
    "    tc_dist_matrix[is.na(tc_dist_matrix)] = 0\n",
    "    \n",
    "    main = substitute(italic(x), list(x=sp))\n",
    "    \n",
    "    par(mar=c(4,5,2,1)); draw_boxplot2(sra, tc_dist_matrix, main, fontsize)    \n",
    "    \n",
    "    #if (sp=='Anolis carolinensis') {\n",
    "    #    break\n",
    "    #}\n",
    "}\n",
    "graphics.off()\n",
    "cat('done.\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Anolis carolinensis \n",
      "Astyanax mexicanus \n",
      "Bos taurus \n",
      "Callithrix jacchus \n",
      "Canis lupus \n",
      "Chinchilla lanigera \n",
      "Danio rerio \n",
      "Gadus morhua \n",
      "Gallus gallus \n",
      "Homo sapiens \n",
      "Macaca mulatta \n",
      "Monodelphis domestica \n",
      "Mus musculus \n",
      "Oreochromis niloticus \n",
      "Ornithorhynchus anatinus \n",
      "Oryctolagus cuniculus \n",
      "Oryzias latipes \n",
      "Ovis aries \n",
      "Rattus norvegicus \n",
      "Sus scrofa \n",
      "Xenopus tropicalis \n",
      "done.\n"
     ]
    }
   ],
   "source": [
    "# predictor of surrogate variables\n",
    "\n",
    "draw_sva_summary2 = function(sva_out, tc, sra, main, fontsize) {\n",
    "    if ((is.null(sva_out))|(class(sva_out)=='try-error')) {\n",
    "        plot(c(0, 1), c(0, 1), ann = F, bty = 'n', type = 'n', xaxt = 'n', yaxt = 'n', main=main)\n",
    "        df = NA\n",
    "    } else {\n",
    "        out = tc_sra_intersect(tc, sra) ; tc = out[['tc']] ; sra = out[['sra']]\n",
    "        out = sort_tc_and_sra(tc, sra) ; tc = out[[\"tc\"]] ; sra = out[[\"sra\"]]\n",
    "        sra$fraction_lost_fastp = 1 - (sra$num_read_fastp / sra$num_read_unfiltered)\n",
    "        sra$fraction_lost_mask = 1 - (sra$num_read_mask / sra$num_read_fastp)\n",
    "        cols = c('tissue','bioproject','lib_selection','layout','instrument','num_read_masked','fraction_lost_fastp',\n",
    "                 'fraction_lost_mask','min_read_len_masked','avg_read_len_masked','max_read_len_masked','mapping_rate_masked')\n",
    "        label_cols = c('organ','BioProject','library selection','library layout','instrument','number of read','% lost, fastp',\n",
    "                       '% lost, misc feature','minimum read length','average read length','maximum read length','mapping rate')\n",
    "        num_sv = sva_out$n.sv\n",
    "        df = data.frame(matrix(NA, num_sv, length(cols)))\n",
    "        colnames(df) = cols\n",
    "        rownames(df) = paste0('SV', 1:nrow(df))\n",
    "        for (i in 1:length(cols)) {\n",
    "            for (j in 1:num_sv) {\n",
    "                if (length(unique(sra[,cols[i]]))==1) {\n",
    "                    df[j,i] = NA\n",
    "                } else {\n",
    "                    df[j,i] = summary(lm(sva_out$sv[,j] ~ sra[,cols[i]]))$adj.r.squared\n",
    "                }\n",
    "            }\n",
    "        }\n",
    "        colnames(df) = label_cols\n",
    "        breaks = seq(0, 1, 0.02)\n",
    "        colors = colorRampPalette(c(\"blue\", \"yellow\", \"red\"))(length(breaks))\n",
    "        df2 = t(df)\n",
    "        df2[df2<0] = 0\n",
    "        aheatmap(df2, color=colors, Rowv=NA, Colv=NA, revC=TRUE, main=main, legend=FALSE,\n",
    "                 breaks=breaks, fontsize=fontsize, labRow=NA)\n",
    "    }\n",
    "    return(df)\n",
    "}\n",
    "\n",
    "fontsize = 6\n",
    "spp = c('Bos taurus','Homo sapiens','Gallus gallus','Macaca mulatta','Canis lupus','Sus scrofa','Ovis aries','Mus musculus','Callithrix jacchus',\n",
    "'Rattus norvegicus','Monodelphis domestica','Oreochromis niloticus','Oryctolagus cuniculus','Danio rerio','Anolis carolinensis','Chinchilla lanigera',\n",
    "'Astyanax mexicanus','Xenopus tropicalis','Oryzias latipes','Gadus morhua','Ornithorhynchus anatinus')\n",
    "spp = spp[order(spp)]\n",
    "\n",
    "pdf('paneled_species_SV_predictor.pdf', height=7, width=7.2, fonts='Helvetica', pointsize=fontsize)\n",
    "#layout_matrix=matrix(c(\n",
    "#    1,2,3,4,5,6,0,\n",
    "#    7,8,9,10,11,12,0,\n",
    "#    13,14,15,16,17,18,0,\n",
    "#    19,20,21,22,23,24,0\n",
    "#    ),\n",
    "#  4, 7, byrow=TRUE\n",
    "#)\n",
    "#layout(layout_matrix)\n",
    "par(mfrow=c(4,6), mar=c(0,0,0,0))\n",
    "\n",
    "for (sp in spp) {\n",
    "    cat(sp, '\\n')\n",
    "    sp_filled = sub(' ', '_', sp)\n",
    "    \n",
    "    files = list.files(dir_sva)\n",
    "    files = files[grep(sp_filled, files)]\n",
    "    file = files[length(files)]\n",
    "    load(paste0(dir_sva, file))\n",
    "    \n",
    "    files = list.files(dir_sra)\n",
    "    file = files[grep(sp_filled, files)]\n",
    "    sra = read.table(paste0(dir_sra, file), sep='\\t', header=TRUE, comment.char='', check.names=FALSE, quote='', stringsAsFactors=FALSE)    \n",
    "    \n",
    "    files = list.files(dir_tc)\n",
    "    file = files[grep(sp_filled, files)]\n",
    "    tc = read.table(paste0(dir_tc, file), sep='\\t', header=TRUE, quote='', stringsAsFactors=FALSE)\n",
    "\n",
    "    sra = add_color_to_sra(sra, selected_tissues)\n",
    "    out = tc_sra_intersect(tc, sra) ; tc = out[['tc']] ; sra = out[['sra']]\n",
    "    out = sort_tc_and_sra(tc, sra) ; tc = out[[\"tc\"]] ; sra = out[[\"sra\"]]\n",
    "    \n",
    "    main = substitute(italic(x), list(x=sp))\n",
    "    \n",
    "    par(mar=rep(0.1,4)); df_r2 = draw_sva_summary2(sva_out, tc, sra, main, fontsize)\n",
    "    \n",
    "    #if (sp=='Anolis carolinensis') {\n",
    "    #    break\n",
    "    #}\n",
    "}\n",
    "graphics.off()\n",
    "cat('done.\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 155,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Anolis carolinensis \n",
      "Astyanax mexicanus \n",
      "Bos taurus \n",
      "Callithrix jacchus \n",
      "Canis lupus \n",
      "Chinchilla lanigera \n",
      "Danio rerio \n",
      "Gadus morhua \n",
      "Gallus gallus \n",
      "Homo sapiens \n",
      "Macaca mulatta \n",
      "Monodelphis domestica \n",
      "Mus musculus \n",
      "Oreochromis niloticus \n",
      "Ornithorhynchus anatinus \n",
      "Oryctolagus cuniculus \n",
      "Oryzias latipes \n",
      "Ovis aries \n",
      "Rattus norvegicus \n",
      "Sus scrofa \n",
      "Xenopus tropicalis \n",
      "done.\n"
     ]
    }
   ],
   "source": [
    "# number of RNA-seq, bioproject, and surrogate variable\n",
    "\n",
    "fontsize = 8\n",
    "spp = c('Bos taurus','Homo sapiens','Gallus gallus','Macaca mulatta','Canis lupus','Sus scrofa','Ovis aries','Mus musculus','Callithrix jacchus',\n",
    "'Rattus norvegicus','Monodelphis domestica','Oreochromis niloticus','Oryctolagus cuniculus','Danio rerio','Anolis carolinensis','Chinchilla lanigera',\n",
    "'Astyanax mexicanus','Xenopus tropicalis','Oryzias latipes','Gadus morhua','Ornithorhynchus anatinus')\n",
    "spp = spp[order(spp)]\n",
    "\n",
    "pdf('paneled_species_num_data.pdf', height=3, width=4.5, fonts='Helvetica', pointsize=fontsize)\n",
    "layout_matrix=matrix(c(1,1,1,1,1,2,2,2,3,3,3), 1, 11, byrow=TRUE)\n",
    "layout(layout_matrix)\n",
    "\n",
    "num_exp = c()\n",
    "num_bioproject = c()\n",
    "num_sv = c()\n",
    "for (sp in spp) {\n",
    "    cat(sp, '\\n')\n",
    "    sp_filled = sub(' ', '_', sp)\n",
    "    \n",
    "    files = list.files(dir_sva)\n",
    "    files = files[grep(sp_filled, files)]\n",
    "    file = files[length(files)]\n",
    "    load(paste0(dir_sva, file))\n",
    "    \n",
    "    files = list.files(dir_sra)\n",
    "    file = files[grep(sp_filled, files)]\n",
    "    sra = read.table(paste0(dir_sra, file), sep='\\t', header=TRUE, comment.char='', check.names=FALSE, quote='', stringsAsFactors=FALSE)    \n",
    "    sra = sra[(sra$exclusion=='no'),]\n",
    "    \n",
    "    num_exp = c(num_exp, nrow(sra))\n",
    "    names(num_exp)[length(num_exp)] = sp\n",
    "    \n",
    "    num_bioproject = c(num_bioproject, length(unique(sra[['bioproject']])))\n",
    "    names(num_bioproject)[length(num_bioproject)] = sp\n",
    "    \n",
    "    if ('n.sv' %in% names(sva_out)) {\n",
    "        num_sv = c(num_sv, sva_out[['n.sv']])\n",
    "    } else {\n",
    "        num_sv = c(num_sv, 0)\n",
    "    }\n",
    "    names(num_sv)[length(num_sv)] = sp\n",
    "}\n",
    "\n",
    "ind = order(num_exp)\n",
    "num_exp = num_exp[ind]\n",
    "num_bioproject = num_bioproject[ind]\n",
    "num_sv = num_sv[ind]\n",
    "\n",
    "par(ps=fontsize, mar=c(4,10,0.1,1))\n",
    "barplot(num_exp, las=1, col='black', xlab='Number of RNA-seq experiments', horiz=TRUE, xlim=c(0,400), font=3, xaxt='n')\n",
    "axis(1)\n",
    "par(ps=fontsize, mar=c(4,0.1,0.1,1))\n",
    "barplot(num_bioproject, las=1, col='black', xlab='Number of BioProjects', yaxt='n', horiz=TRUE, xlim=c(0,40))\n",
    "barplot(num_sv, las=1, col='black', xlab='Number of sorrogate variables', yaxt='n', horiz=TRUE, xlim=c(0,30))\n",
    "\n",
    "\n",
    "graphics.off()\n",
    "cat('done.\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "metadata": {},
   "outputs": [],
   "source": [
    "graphics.off()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R-Transcriptome",
   "language": "R",
   "name": "r-transcriptome"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.4.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
